File: daemon/src/agent-plugins/registry.ts
/**
 * Agent plugin registry.
 * Resolves concrete AgentPlugin implementations by CLI name.
 */

import type { AgentPlugin } from "../services/spawn.js";
import { createClaudePlugin } from "./claude.js";
import { createCursorPlugin } from "./cursor.js";
import { createGeminiPlugin } from "./gemini.js";
import { createOpencodePlugin } from "./opencode.js";

export const PLUGIN_MAP = {
  claude: createClaudePlugin,
  cursor: createCursorPlugin,
  opencode: createOpencodePlugin,
  gemini: createGeminiPlugin,
} as const satisfies Record<string, () => AgentPlugin>;

export type CliId = keyof typeof PLUGIN_MAP;

export const SUPPORTED_CLIS = Object.keys(PLUGIN_MAP) as CliId[];

/**
 * Resolve a concrete AgentPlugin for the given CLI identifier.
 * @throws Error if the CLI is not recognized.
 */
export function resolvePlugin(cli: CliId): AgentPlugin {
  if (!(cli in PLUGIN_MAP)) {
    throw new Error(`Unknown CLI: ${String(cli)}`);
  }
  return PLUGIN_MAP[cli]();
}

File: rust/vst-agents/src/registry.rs
//! Agent plugin registry — ports `agent-plugins/registry.ts`.
//!
//! Resolves a concrete [`AgentPlugin`] implementation by CLI id. Per
//! `AGENTS.md` § Agent plugin + arch Gotcha #2 / System Boundaries, callers
//! dispatch through the returned trait object — they never match on a `CliId`
//! enum after resolving.

use crate::agy::create_agy_plugin;
use crate::claude::create_claude_plugin;
use crate::cursor::create_cursor_plugin;
use crate::opencode::create_opencode_plugin;
use crate::plugin::AgentPlugin;
use vst_types::CliId;

/// All supported CLI ids, in registry order. Mirrors `SUPPORTED_CLIS`.
pub const SUPPORTED_CLIS: [CliId; 4] = [CliId::Claude, CliId::Cursor, CliId::Opencode, CliId::Agy];

/// Resolve a concrete [`AgentPlugin`] for the given CLI identifier.
///
/// Returns a fresh plugin instance per call, mirroring the TS
/// `PLUGIN_MAP[cli]()` factories. Panics (programmer error) on an unknown CLI.
pub fn resolve_plugin(cli: CliId) -> Box<dyn AgentPlugin> {
    match cli {
        CliId::Claude => Box::new(create_claude_plugin()),
        CliId::Cursor => Box::new(create_cursor_plugin()),
        CliId::Opencode => Box::new(create_opencode_plugin()),
        CliId::Agy => Box::new(create_agy_plugin()),
    }
}
